File: leetcode/jumpGame.js
// Leetcode 55: Jump Game

var canJump = function(nums) {
  let len = nums.length;
  let lastGoodIndex = len - 1;
  for (let i = len - 2; i >= 0; i--) {
      if (i + nums[i] >= lastGoodIndex) {
          lastGoodIndex = i;
      }
  }
  return lastGoodIndex === 0;
};

console.log(canJump([3,2,1,0,4]))

File: leetcode/jump_game.py
# Leetcode 55. Jump Game

# Approach: Greedy | Time - O(n) | Space - O(1)
from typing import List
class Solution:
  def canJump(self, nums: List[int]) -> int:
    n = len(nums)
    goal = n - 1

    for i in range(n - 1, -1, -1):
      if nums[i] + i >= goal:
        goal = i
    
    return goal == 0
  

if __name__ == "__main__":
  sol = Solution()
  print(sol.canJump([3,2,1,0,4]))